src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -238,27 +226,131 @@ void MulticopterPositionControl::Run()
+						_setpoint.y += local_pos.delta_xy[1];
+					}
+				}
 
-				if (constraints.reset_integral) {
-					_control.resetIntegral();
+				if (local_pos.z_reset_counter != _z_reset_counter) {
+					if (PX4_ISFINITE(_setpoint.z)) {
+						_setpoint.z += local_pos.delta_z;
+					}
 				}
+
+				if (local_pos.heading_reset_counter != _heading_reset_counter) {
+					if (PX4_ISFINITE(_setpoint.yaw)) {
+						_setpoint.yaw += local_pos.delta_heading;
+					}
+				}
+			}
+
+			// update vehicle constraints and handle smooth takeoff
+			_vehicle_constraints_sub.update(&_vehicle_constraints);
+
+			// fix to prevent the takeoff ramp to ramp to a too high value or get stuck because of NAN
+			// TODO: this should get obsolete once the takeoff limiting moves into the flight tasks
+			if (!PX4_ISFINITE(_vehicle_constraints.speed_up) || (_vehicle_constraints.speed_up > _param_mpc_z_vel_max_up.get())) {
+				_vehicle_constraints.speed_up = _param_mpc_z_vel_max_up.get();
+			}
+
+			// handle smooth takeoff
+			_takeoff.updateTakeoffState(_control_mode.flag_armed, _vehicle_land_detected.landed, _vehicle_constraints.want_takeoff,
+						    _vehicle_constraints.speed_up, false, time_stamp_now);
+
+			const bool not_taken_off = (_takeoff.getTakeoffState() < TakeoffState::rampup);
+			const bool flying = (_takeoff.getTakeoffState() >= TakeoffState::flight);
+			const bool flying_but_ground_contact = (flying && _vehicle_land_detected.ground_contact);
+
+			if (not_taken_off || flying_but_ground_contact) {
+				// we are not flying yet and need to avoid any corrections
+				reset_setpoint_to_nan(_setpoint);
+				Vector3f(0.f, 0.f, 100.f).copyTo(_setpoint.acceleration); // High downwards acceleration to make sure there's no thrust
+
+				// prevent any integrator windup
+				_control.resetIntegral();
 			}
 
+			// limit tilt during takeoff ramupup
+			if (_takeoff.getTakeoffState() < TakeoffState::flight) {
+				_control.setTiltLimit(math::radians(_param_mpc_tiltmax_lnd.get()));
+
+			} else {
+				_control.setTiltLimit(math::radians(_param_mpc_tiltmax_air.get()));
+			}
+
+			const float speed_up = _takeoff.updateRamp(dt, _vehicle_constraints.speed_up);
+			const float speed_down = PX4_ISFINITE(_vehicle_constraints.speed_down) ? _vehicle_constraints.speed_down :
+						 _param_mpc_z_vel_max_dn.get();
+			const float speed_horizontal = PX4_ISFINITE(_vehicle_constraints.speed_xy) ? _vehicle_constraints.speed_xy :
+						       _param_mpc_xy_vel_max.get();
+
+			// Allow ramping from zero thrust on takeoff
+			const float minimum_thrust = flying ? _param_mpc_thr_min.get() : 0.f;
+
+			_control.setThrustLimits(minimum_thrust, _param_mpc_thr_max.get());
+
+			_control.setVelocityLimits(
+				math::constrain(speed_horizontal, 0.f, _param_mpc_xy_vel_max.get()),
+				math::constrain(speed_up, 0.f, _param_mpc_z_vel_max_up.get()),
+				math::constrain(speed_down, 0.f, _param_mpc_z_vel_max_dn.get()));
+
+			_control.setInputSetpoint(_setpoint);
+
+			// update states
```

Comment:
Preparation to test without this blending right? Such that we can either live completely without or improve the estimate if (still) necessary.

Reply:
This was just keeping it contained, but similar to my previous comment (https://github.com/PX4/PX4-Autopilot/pull/16869#discussion_r579827274), I really don't like when we can't see the data that's actually being used.

It's very likely minor, but the plots on flight review are still lying to us.

Reply:
Yes, I'm not sure just removing this blending will not bring back any issue but we should consider it and like you suggest fix the logged state in use and not do something internal and local.